### PR TITLE
Export to Office 2007

### PR DESCRIPTION
**Pull request by Siedlerchr**
First fix for #1750.
I will continue reworking some other fields, too and will add some more tests.

- [X] Change in CHANGELOG.md described
- [X] Tests created for changes
- [X] Manually tested changed features in running JabRef
- [X] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest7.bib [hunk 1]
```diff
@@ -0,0 +1,15 @@
+% Encoding: UTF-8
+
+@InBook{Orlowski;2011,
```
Comment (stefan-kolb) [inline src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest7.bib]: Please name the test files after what they do and minimize them, not like the Stupro did just enumerating them :smile:

**Comment by Siedlerchr** [comment]
I would like to merge this in for 3,6 (No Localization changes), after I rebase

File: CHANGELOG.md [hunk 2]
```diff
@@ -83,8 +83,9 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Fixed [#1531](https://github.com/JabRef/jabref/issues/1531): `\relax` can be used for abbreviation of author names
 - Fixed [#1771](https://github.com/JabRef/jabref/issues/1771): Show all supported import types as default
 - Fixed [#1804](https://github.com/JabRef/jabref/issues/1804): Integrity check no longer removes URL field by mistake
-
- 
+- Fixed [#1750](https://github.com/JabRef/jabref/issues/1750): BibLaTeX `date` field is now correctly exported as `year` in MS-Office 2007 xml format
+- Fixed: LaTeX characters in author names are now converted to Unicode before export in MS-Office 2007 xml format
+- Fixed: `Volume`, `Journaltitle`,  `Issue` and `number`(for Patents) fields are now exported correctly in MS-Office 2007 xml format
```
Comment (koppor) [inline CHANGELOG.md]: all field names in lower case. Space after `number`. Patents with lower case "p"

File: src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java [hunk 3]
```diff
@@ -77,12 +81,26 @@ public static BibEntry convert(MSBibEntry entry) {
             fieldValues.put(MSBIB_PREFIX + "accessed", entry.dateAccessed);
         }
 
+        if(entry.journalName != null)
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java]: Don't forget to add the space after the `if` and move the bracket in the same line.

File: src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java [hunk 4]
```diff
@@ -77,12 +81,26 @@ public static BibEntry convert(MSBibEntry entry) {
...
+        }
+
+
         // set all fields
         result.setField(fieldValues);
 
         return result;
     }
 
+
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java]: Why another emtpy line?

File: src/test/resources/net/sf/jabref/logic/importer/fileformat/MsBibImporterTest3.bib [hunk 5]
```diff
@@ -1,10 +1,12 @@
+% Encoding: UTF-8
+
 @book{raey,
   author = {5},
-  journal = {Wirtschaftsinformatik},
+  journaltitle = {Wirtschaftsinformatik},
```
Comment (koppor) [inline src/test/resources/net/sf/jabref/logic/importer/fileformat/MsBibImporterTest3.bib]: I don't see biblatex mode activated for that file (respective `jabref-comment` missing). How can this test pass? Because of the autodetection of the bib file mode?

File: src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest1.xml [hunk 6]
```diff
@@ -1,67 +1,34 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <b:Sources xmlns:b="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" xmlns="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" SelectedStyle="">
 <b:Source>
-<b:SourceType>Report</b:SourceType>
```
Comment (koppor) [inline src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest1.xml]: Why is this test file changed? Why isn't there generated a new file? Was the old file an invalid MsBib file?

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 7]
```diff
@@ -103,19 +111,32 @@ public static MSBibEntryType getMSBibEntryType(String bibtexType) {
...
+     * @return Returns 0 for English
+     */
+    @Deprecated
     public static int getLCID(String language) {
         // TODO: add language to LCID mapping
         // 0 is English
         return 0;
     }
 
     // http://www.microsoft.com/globaldev/reference/lcid-all.mspx
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Integrate that in the method body (or as real comment in the method comment)

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 8]
```diff
@@ -103,19 +111,32 @@ public static MSBibEntryType getMSBibEntryType(String bibtexType) {
...
         entryTypeMapping.put("patent", MSBibEntryType.Patent);
         entryTypeMapping.put("misc", MSBibEntryType.Misc);
-        entryTypeMapping.put("electronic", MSBibEntryType.Misc);
+        entryTypeMapping.put("electronic", MSBibEntryType.ElectronicSource);
+        entryTypeMapping.put("online", MSBibEntryType.InternetSite);
 
         return entryTypeMapping.getOrDefault(bibtexType, MSBibEntryType.Misc);
     }
 
     // http://www.microsoft.com/globaldev/reference/lcid-all.mspx
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: see blow

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 9]
```diff
@@ -52,34 +58,36 @@
...
 
         Map<String, String> entryTypeMapping = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
 
-        entryTypeMapping.put("Book", BibtexEntryTypes.BOOK.getName());
-        entryTypeMapping.put("BookSection", BibtexEntryTypes.INBOOK.getName());
-        entryTypeMapping.put("JournalArticle", BibtexEntryTypes.ARTICLE.getName());
-        entryTypeMapping.put("ArticleInAPeriodical", BibtexEntryTypes.ARTICLE.getName());
-        entryTypeMapping.put("ConferenceProceedings", BibtexEntryTypes.CONFERENCE.getName());
-        entryTypeMapping.put("Report", BibtexEntryTypes.TECHREPORT.getName());
+        entryTypeMapping.put("Book", BibLatexEntryTypes.BOOK.getName());
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Why isn't there a distinction between bibtex and biblatex? - Currently, the importers all do bibtex and the conversion to biblatex should be done afterwards (currently not implemented). For me, it would be also OK to parameterize the importers to generate bibtex or biblatex.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 10]
```diff
@@ -4,12 +4,19 @@
...
 
 import net.sf.jabref.model.entry.BibEntry;
+import net.sf.jabref.model.entry.BibLatexEntryTypes;
 import net.sf.jabref.model.entry.BibtexEntryTypes;
 import net.sf.jabref.model.entry.FieldName;
 
 import com.google.common.collect.HashBiMap;
 
+/**
+ * Mapping between Msbib and Bibtex
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Why not "biblatex"? See below.

File: src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java (hunk 4 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java]: Maybe a result of the rebasing. Will fix it.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java (hunk 9 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: The thing is: Patent and InternetSite (Online) would both get mapped to misc, as this are non avaiable entry types in bibte
Therefore I chose biblatex to make it consistent.

File: src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest1.xml (hunk 6 above)
Comment (Siedlerchr) [inline src/test/resources/net/sf/jabref/logic/exporter/MsBibExportFormatTest1.xml]: I removed some fields to have shorter test files. It does not make sense if every field is tested in every file, Makes it only harder to spot errors.

File: src/test/resources/net/sf/jabref/logic/importer/fileformat/MsBibImporterTest3.bib (hunk 5 above)
Comment (Siedlerchr) [inline src/test/resources/net/sf/jabref/logic/importer/fileformat/MsBibImporterTest3.bib]: Just maps the fields from the conversion HashMap from MsBib to the BibLaTex fields, so the test doesn't check for the bibmode.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java (hunk 7 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Isn't `@Deprecated` for methods that eventually should be removed, rather than for methods that are still not properly implemented? I think that with increased BibLatex support we will add support for `language` and `lang-id`. If they are in any way related to the lcid I have no idea about though...

File: src/main/java/net/sf/jabref/logic/msbib/MSBibEntry.java [hunk 11]
```diff
@@ -129,19 +143,26 @@ private void populateFromXml(Element entry) {
...
         internetSiteTitle = getXmlElementTextContent("InternetSiteTitle", entry);
-        String month = getXmlElementTextContent("MonthAccessed", entry);
-        String day = getXmlElementTextContent("DayAccessed", entry);
-        String year = getXmlElementTextContent("YearAccessed", entry);
+
+        String monthAccessed = getXmlElementTextContent("MonthAccessed", entry);
+        String dayAccessed = getXmlElementTextContent("DayAccessed", entry);
+        String yearAccessed = getXmlElementTextContent("YearAccessed", entry);
+
         dateAccessed = "";
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibEntry.java]: StringBuilder/StringBuffer?

File: src/main/java/net/sf/jabref/model/entry/BibEntry.java [hunk 12]
```diff
@@ -207,15 +207,15 @@ public void setType(EntryType type) {
...
      */
-    @Deprecated //Use getFieldOptional instead
+    @Deprecated
     public String getField(String name) {
         return fields.get(toLowerCase(name));
     }
 
     /**
-     * Returns the contents of the given field as an Optional.
+     * Returns the contents of the given field as an Optional. The optional may contain null
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/model/entry/BibEntry.java]: No, it may not. The Optional will be empty in case it was supposed to contain null.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java (hunk 7 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Bah, removed the comment by mistake... Please see the email...

Basically: should it be `@Deprecated` if it is a method that should be improved rather than removed?

**Comment by oscargus** [comment]
Looks good! Just some minor comments.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java (hunk 7 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Hm, yeah, I was not sure how to do it best, but I wanted to give a clear sign that the method atm is not fully impleemented

**Comment by Siedlerchr** [comment]
I will fix it this afternoon/evening. Regarding the language mapping, there is not yet a simple solution, as we would have to integrate the whole table of LCID: 


https://msdn.microsoft.com/de-de/goglobal/bb964664.aspx
https://support.office.com/en-us/article/Locale-identification-numbers-for-language-specific-files-16102229-37c5-4ad5-92b1-0438f6304114?ui=en-US&rs=en-US&ad=US

File: src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java [hunk 13]
```diff
@@ -77,12 +81,26 @@ public static BibEntry convert(MSBibEntry entry) {
             fieldValues.put(MSBIB_PREFIX + "accessed", entry.dateAccessed);
         }
 
+        if(entry.journalName != null)
+        {
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/msbib/BibTeXConverter.java]: { in line above

File: src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java [hunk 14]
```diff
@@ -2,18 +2,20 @@
...
 import net.sf.jabref.logic.mods.PersonName;
 import net.sf.jabref.model.entry.BibEntry;
 import net.sf.jabref.model.entry.FieldName;
 
 public class MSBibConverter {
+
     private static final String MSBIB_PREFIX = "msbib-";
     private static final String BIBTEX_PREFIX = "BIBTEX_";
 
+
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java]: Empty line

File: src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java [hunk 15]
```diff
@@ -135,10 +127,6 @@ public static MSBibEntry convert(BibEntry entry) {
     }
 
     private static String removeLaTeX(String text) {
-        // TODO: just use latex free version everywhere in the future
-        String result = new RemoveBrackets().format(text);
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java]: Why did you remove the bracketsformatter here? I added it so we dont end up wit `{ACM} Cooperation` in Word bibliographies. As long as we don't have a LaTeX-free BibEntry version we should keep it that way?

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 16]
```diff
@@ -103,19 +111,32 @@ public static MSBibEntryType getMSBibEntryType(String bibtexType) {
...
         return entryTypeMapping.getOrDefault(bibtexType, MSBibEntryType.Misc);
     }
 
     // http://www.microsoft.com/globaldev/reference/lcid-all.mspx
+    /**
+     *  Only English is supported
+     * @param language
+     * @return Returns 0 for English
+     */
+    @Deprecated
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: Just use comments there for now and keep it the way it is.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java [hunk 17]
```diff
@@ -103,19 +111,32 @@ public static MSBibEntryType getMSBibEntryType(String bibtexType) {
...
         return 0;
     }
 
     // http://www.microsoft.com/globaldev/reference/lcid-all.mspx
+    /**
+     * Only English is supported
+     * @param language
+     * @return Returns english
+     */
+    @Deprecated
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibMapping.java]: See above.

File: src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java (hunk 15 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java]: LatexToUnicode removes the brackets as well.

Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/msbib/MSBibConverter.java]: The Bracketsformatter broke things up when run before the Unicode formatter. E.g. the unicode was not correclty converted.

**Comment by Siedlerchr** [comment]
Address comments as far as possible
